src/app/profiles.rs

```diff
@@ -0,0 +1,235 @@
+            state: State::Disconnected,
+            service: self.service.clone(),
+            backoff: self.backoff,
+        })
+    }
+}
+
+// === impl Rx ===
+
+impl<T> Stream for Rx<T>
+where
+    T: HttpService<RequestBody = BoxBody> + Clone,
+    T::ResponseBody: Body<Data = Data>,
+    T::Error: fmt::Debug,
+{
+    type Item = Vec<(profiles::RequestMatch, profiles::Route)>;
+    type Error = profiles::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let service = match self.service {
+            Some(ref s) => s,
+            None => return Ok(Async::Ready(Some(Vec::new()))),
+        };
+
+        loop {
+            self.state = match self.state {
+                State::Disconnected => {
+                    let mut client = api::client::Destination::new(service.clone());
+                    let req = api::GetDestination {
+                        scheme: "k8s".to_owned(),
+                        path: self.dst.clone(),
+                    };
+                    debug!("disconnected; getting profile: {:?}", req);
+                    let rspf = client.get_profile(grpc::Request::new(req));
+                    State::Waiting(rspf)
+                }
+                State::Waiting(ref mut f) => match f.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(rsp)) => {
+                        debug!("response received");
+                        State::Streaming(rsp.into_inner())
+                    }
+                    Err(e) => {
+                        warn!("error fetching profile for {}: {:?}", self.dst, e);
+                        State::Backoff(Delay::new(clock::now() + self.backoff))
+                    }
+                },
+                State::Streaming(ref mut s) => match s.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(Some(profile))) => {
+                        debug!("profile received: {:?}", profile);
+                        let routes = Self::convert_routes(profile);
+                        return Ok(Async::Ready(Some(routes)));
+                    }
+                    Ok(Async::Ready(None)) => {
+                        debug!("profile stream ended");
+                        State::Backoff(Delay::new(clock::now() + self.backoff))
+                    }
+                    Err(e) => {
+                        warn!("profile stream failed: {:?}", e);
+                        State::Backoff(Delay::new(clock::now() + self.backoff))
+                    }
+                },
+                State::Backoff(ref mut f) => match f.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Err(_) | Ok(Async::Ready(())) => State::Disconnected,
+                },
+            };
+        }
+    }
+}
+
+impl<T: HttpService> Rx<T> {
+    fn convert_routes(
```

Comment:
Nit/TIOLI: It's a little strange to me that these functions are all defined on `Rx`, since none of them actually involve an `Rx` or any of it's type parameters. If they're only intended to be used within the future impl, perhaps they should just be free functions defined inside `poll`? Not a big deal though.